client/service/src/builder.rs

```diff
@@ -51,7 +51,71 @@ use sysinfo::{get_current_pid, ProcessExt, System, SystemExt};
 use sc_telemetry::{telemetry, SUBSTRATE_INFO};
 use sp_transaction_pool::{TransactionPool, TransactionPoolMaintainer};
 use sp_blockchain;
-use grafana_data_source::{self, record_metrics};
+use prometheus_exporter::{create_gauge, Gauge, U64, F64, Registry, PrometheusError};
+
+struct ServiceMetrics {
+	finality_height: Gauge<U64>,
+	best_height: Gauge<U64>,
+	peers_num: Gauge<U64>,
+	tx_count: Gauge<U64>,
+	node_memory: Gauge<U64>,
+	node_cpu: Gauge<F64>,
+	node_download: Gauge<U64>,
+	node_upload: Gauge<U64>,
+	sync_target: Gauge<U64>,
+}
+
+impl ServiceMetrics {
+	fn register(registry: &Registry) -> Result<Self, PrometheusError> {
+		Ok(Self {
+			finality_height: create_gauge(
+				"finality_block_height_number",
```

Comment:
Instead of having two distinct metrics `finality_block_height_number` and `best_block_height_number`, one could also have a single metric `block_height_number` with two label instances `finality` and `best`. With this format one can easier run queries against the two in combination and display them in a single graph within the Prometheus web interface. (Grafana can display them in a single graph anyways.)

What do you think?

Reply:
Are there two associations? For example
In the 10th best block, the finality block must be the 5th ..?

If there is finality block information in the best block, it is better to integrate it using gaugevec.

Reply:
@nodebreaker0-0 I might not have explained that very well. Let me add some examples:

Instead of doing

```
finality_block_height_number{target="foo"} 11
best_block_height_number{target="foo"} 20
```

we could as well do

```
block_height_number{status="finalized", target="foo"} 11
block_height_number{status="best", target="foo"} 20
```

Reply:
Good feedback, I think you can use vec
https://github.com/nodebreaker0-0/substrate/pull/7

I think we can discuss at.